lib/audit/auditor.go

```diff
@@ -0,0 +1,362 @@
+	r, repoPath, err := cloneToTempDir(repoURL, branch)
+	if err != nil {
+		return nil, err
+	}
+
+	erc := initErrorReportingClient(gcpProjectID)
+	logClient := initLogClient(gcpProjectID)
+
+	serverContext := ServerContext{
+		RepoURL:              repoURL,
+		RepoBranch:           branch,
+		ThinManifestDirPath:  path,
+		Repo:                 r,
+		RepoPath:             repoPath,
+		ErrorReportingClient: erc,
+		LogClient:            logClient,
+	}
+
+	return &serverContext, nil
+}
+
+// initLogClient creates a logging client that performs better logging than the
+// default behavior on GCP Stackdriver. For instance, logs sent with this client
+// are not split up over newlines, and also the severity levels are actually
+// understood by Stackdriver.
+func initLogClient(projectID string) *logging.Client {
+
+	ctx := context.Background()
+
+	// Creates a client.
+	client, err := logging.NewClient(ctx, projectID)
+	if err != nil {
+		klog.Fatalf("Failed to create client: %v", err)
+	}
+
+	return client
+}
+
+func initErrorReportingClient(projectID string) *errorreporting.Client {
+
+	ctx := context.Background()
+
+	erc, err := errorreporting.NewClient(ctx, projectID, errorreporting.Config{
+		ServiceName: "cip-auditor",
+		OnError: func(err error) {
+			klog.Errorf("Could not log error: %v", err)
+		},
+	})
+	if err != nil {
+		klog.Fatalf("Failed to create errorreporting client: %v", err)
+	}
+
+	return erc
+}
+
+// Auditor runs an HTTP server.
+func Auditor(gcpProjectID, repoURL, branch, path string) {
+	klog.Info("Starting Auditor")
+	serverContext, err := initServerContext(gcpProjectID, repoURL, branch, path)
+	if err != nil {
+		klog.Exitln(err)
+	}
+
+	klog.Infoln(serverContext)
+	klog.Infoln(serverContext.Repo)
+
+	// nolint[errcheck]
+	defer serverContext.LogClient.Close()
+	// nolint[errcheck]
+	defer serverContext.ErrorReportingClient.Close()
+
+	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		serverContext.Audit(w, r)
+	})
+	// Determine port for HTTP service.
+	port := os.Getenv("PORT")
+	if port == "" {
+		port = "8080"
+		klog.Infof("Defaulting to port %s", port)
+	}
+	// Start HTTP server.
+	klog.Infof("Listening on port %s", port)
+	klog.Fatal(http.ListenAndServe(fmt.Sprintf(":%s", port), nil))
+}
+
+func cloneToTempDir(
+	repoURL fmt.Stringer,
+	branch string,
+) (*git.Repository, string, error) {
+	tdir, err := ioutil.TempDir("", "k8s.io-")
+	if err != nil {
+		return nil, "", err
+	}
+
+	r, err := git.PlainClone(tdir, false, &git.CloneOptions{
+		URL:           repoURL.String(),
+		ReferenceName: (plumbing.ReferenceName)("refs/heads/" + branch),
+	})
+	if err != nil {
+		return nil, "", err
+	}
+
+	sha, err := getHeadSha(r)
+	if err == nil {
+		klog.Infof("updated %v to revision %v", tdir, sha)
+	}
+
+	return r, tdir, nil
+}
+
+// Check if the given path already has a git repository. If it does, just use
+// that. Otherwise, clone a repo to a new temporary path.
+func cloneOrPull(repoURL fmt.Stringer,
```

Comment:
I'm worried aboutthe sorts of corner-cases git-sync experiences here.  I wonder if it would be simpler and more successful to just do a clone with --depth=1 on every alert?

Reply:
It would be simpler. Re-reading this code, it looks like my treatment of ErrNonFastForwardUpdate is wrong (I should always fetch and then reset-hard to the latest branch HEAD, because if I do hit a ErrNonFastForwardUpdate error the library will not actually update the branch HEAD).

I will just clone every time and take care to delete the tmp directory after I'm done reading the manifests. ACK